### PR TITLE
Ingest MMR with Proof

### PR DESCRIPTION
**Pull request by Wizdave97**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -14,6 +14,191 @@
...
+        {
+            return Err(BeefyClientError::InvalidMmrUpdate);
+        }
+
+        if current_authority_set.id != validator_set_id && next_authority_set.id != validator_set_id
+        {
+            return Err(BeefyClientError::InvalidMmrUpdate);
+        }
+
+        // Beefy validators sign the kekak256 hash of the scale encoded commitment
```
Comment (seunlanlege) [inline src/lib.rs:62]:
```suggestion
        // Beefy validators sign the keccak256 hash of the scale encoded commitment
```

File: src/lib.rs [hunk 2]
```diff
@@ -14,6 +14,191 @@
...
+        let mmr_root_vec = mmr_update
+            .signed_commitment
+            .commitment
+            .payload
+            .get_raw(&MMR_ROOT_ID)
+            .ok_or_else(|| BeefyClientError::InvalidMmrUpdate)?
+            .clone();
+        if mmr_root_vec.len() != HASH_LENGTH {
+            return Err(BeefyClientError::InvalidRootHash);
+        }
```
Comment (seunlanlege) [inline src/lib.rs:162]: this check should be done much earlier

**Comment by seunlanlege** [comment]
Great work overall, but i think the light client should be a struct that holds a generic type `Store` instead

File: src/lib.rs [hunk 3]
```diff
@@ -14,6 +14,200 @@
...
+use rs_merkle::MerkleProof;
+use sp_core::{ByteArray, H256};
+use sp_core_hashing::keccak_256;
+use sp_io::crypto;
+use sp_runtime::traits::Convert;
+
+use sp_std::prelude::*;
+
+pub struct BeefyLightClient<Store: StorageRead + StorageWrite> {
+    _store: Store,
```
Comment (seunlanlege) [inline src/lib.rs:38]: might as well use phantomdata

File: src/lib.rs [hunk 4]
```diff
@@ -14,6 +14,200 @@
...
+        let node =
+            pallet_mmr_primitives::DataOrHash::Data(encodable_opaque_leaf.into_opaque_leaf());
+        match pallet_mmr::verify_leaf_proof::<sp_runtime::traits::Keccak256, _>(
+            mmr_root_hash.into(),
+            node,
+            proof,
+        ) {
+            Err(_) => return Err(BeefyClientError::InvalidMmrProof),
+            _ => {}
+        }
```
Comment (seunlanlege) [inline src/lib.rs:185]: use map_err

File: src/lib.rs [hunk 5]
```diff
@@ -14,6 +14,166 @@
...
+            mmr_root_hash.into(),
+            node,
+            proof,
+        )
+        .map_err(|_| BeefyClientError::InvalidMmrProof)?;
+
+        <Store as StorageWrite>::set_latest_height(
+            mmr_update.signed_commitment.commitment.block_number,
+        )?;
+        <Store as StorageWrite>::set_latest_mmr_root_hash(mmr_root_hash.into())?;
```
Comment (seunlanlege) [inline src/lib.rs:156]: we definitely want these two calls to happen as a transaction, so maybe batch them into 1 struct with a single method.

File: src/lib.rs [hunk 6]
```diff
@@ -14,6 +14,166 @@
...
+
+        if authorities_changed {
+            <Store as StorageWrite>::set_current_authority_set(next_authority_set)?;
+            <Store as StorageWrite>::set_next_authority_set(
+                mmr_update
+                    .latest_mmr_leaf_with_index
+                    .leaf
+                    .beefy_next_authority_set
+                    .clone(),
+            )?;
```
Comment (seunlanlege) [inline src/lib.rs:166]: same as this

File: src/lib.rs [hunk 7]
```diff
@@ -14,6 +14,192 @@
...
+                    root
+                } else {
+                    return Err(BeefyClientError::InvalidRootHash);
+                }
+            } else {
+                return Err(BeefyClientError::InvalidMmrUpdate);
+            }
+        };
+
+        let mut mmr_root_hash = [0u8; 32];
```
Comment (seunlanlege) [inline src/lib.rs:90]: You can do H256::from_slice

File: src/lib.rs [hunk 8]
```diff
@@ -14,6 +14,192 @@
...
+        let authority_leaves = authority_addresses_and_indices
+            .into_iter()
+            .map(|x| keccak_256(&x.1))
+            .collect::<Vec<_>>();
+
+        let authorities_merkle_proof =
+            rs_merkle::MerkleProof::<KeccakHasher>::new(mmr_update.authority_proof);
+
+        // Verify mmr_update.authority_proof against store root hash
+        if current_authority_set.id == validator_set_id {
```
Comment (seunlanlege) [inline src/lib.rs:138]: Let's use a match statement

File: src/tests.rs [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+            let authority_proof = tree.proof(&signature_indices);
+
+            let mmr_update = MmrUpdateProof {
+                signed_commitment: SignedCommitment {
+                    commitment: signed_commitment.commitment,
+                    signatures,
+                },
+                latest_mmr_leaf_with_index: MmrLeafWithIndex {
+                    leaf: latest_leaf,
+                    index: leaf_index,
```
Comment (seunlanlege) [inline src/tests.rs:173]: Let's remove this, we can infer leaf index from the block number in the MMR leaf

File: src/primitives.rs [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+pub const HASH_LENGTH: usize = 32;
+pub type TSignature = [u8; 65];
+pub type Hash = [u8; 32];
+
+#[derive(Clone, sp_std::fmt::Debug, PartialEq, Eq, Encode, Decode)]
+pub struct SignedCommitment {
+    pub commitment: beefy_primitives::Commitment<u32>,
+    pub signatures: Vec<Option<TSignature>>,
```
Comment (seunlanlege) [inline src/primitives.rs:20]: Should be `SignatureWithAuthorityIndex`

File: src/tests.rs [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+                },
+                mmr_proof,
+                authority_proof: authority_proof.proof_hashes().to_vec(),
+            };
+
+            assert_eq!(
+                beef_light_client.ingest_mmr_root_with_proof(mmr_update),
+                Ok(())
+            );
```
Comment (seunlanlege) [inline src/tests.rs:182]: Let's also make asserts on the new mmr state.

**Review by seunlanlege — CHANGES_REQUESTED** [review]

File: src/lib.rs [hunk 12]
```diff
@@ -14,6 +14,182 @@
...
+                    root
+                } else {
+                    return Err(BeefyClientError::InvalidRootHash);
+                }
+            } else {
+                return Err(BeefyClientError::InvalidMmrUpdate);
+            }
+        };
+
+        let mmr_root_hash = H256::from_slice(&*mmr_root_vec);
```
Comment (seunlanlege) [inline src/lib.rs:98]:
```suggestion
        // will panic if `mmr_root_vec.len() < 32` 
        let mmr_root_hash = H256::from_slice(&*mmr_root_vec);
```

Comment (Wizdave97) [reply src/lib.rs:98]: I have a check that exits the function if it's less than 32 bytes before this line

Comment (seunlanlege) [reply src/lib.rs:98]: oh ok nice,

**Review by seunlanlege — APPROVED** [review]

**Pull request closed without merging**